lib/galaxy/datatypes/proteomics.py

```diff
@@ -0,0 +1,405 @@
+
+    def sniff(self, filename):
+        mgf_begin_ions = "BEGIN IONS"
+        max_lines = 100
+
+        with open(filename) as handle:
+            for i, line in enumerate(handle):
+                line = line.rstrip()
+                if line == mgf_begin_ions:
+                    return True
+                if i > max_lines:
+                    return False
+
+
+class MascotDat(Text):
+    """Mascot search results """
+    file_ext = "mascotdat"
+
+    def set_peek(self, dataset, is_multi_byte=False):
+        """Set the peek and blurb text"""
+        if not dataset.dataset.purged:
+            dataset.peek = data.get_file_peek(dataset.file_name, is_multi_byte=is_multi_byte)
+            dataset.blurb = 'mascotdat Mascot Search Results'
+        else:
+            dataset.peek = 'file does not exist'
+            dataset.blurb = 'file purged from disk'
+
+    def sniff(self, filename):
+        mime_version = "MIME-Version: 1.0 (Generated by Mascot version 1.0)"
+        max_lines = 10
+
+        with open(filename) as handle:
+            for i, line in enumerate(handle):
+                line = line.rstrip()
+                if line == mime_version:
+                    return True
+                if i > max_lines:
+                    return False
+
+
+class RAW(Binary):
+    """Class describing a Thermo Finnigan binary RAW file"""
+    file_ext = "raw"
+
+    def sniff(self, filename):
+        # Thermo Finnigan RAW format is proprietary and hence not well documented.
+        # Files start with 2 bytes that seem to differ followed by F\0i\0n\0n\0i\0g\0a\0n
+        # This combination represents 17 bytes, but to play safe we read 20 bytes from
+        # the start of the file.
+        try:
+            header = open(filename).read(20)
+            hexheader = binascii.b2a_hex(header)
+            finnigan = binascii.hexlify('F\0i\0n\0n\0i\0g\0a\0n')
+            if hexheader.find(finnigan) != -1:
+                return True
+            return False
+        except:
+            return False
+
+    def set_peek(self, dataset, is_multi_byte=False):
+        if not dataset.dataset.purged:
+            dataset.peek = "Thermo Finnigan RAW file"
+            dataset.blurb = data.nice_size(dataset.get_size())
+        else:
+            dataset.peek = 'file does not exist'
+            dataset.blurb = 'file purged from disk'
+
+    def display_peek(self, dataset):
+        try:
+            return dataset.peek
+        except:
+            return "Thermo Finnigan RAW file (%s)" % (data.nice_size(dataset.get_size()))
+
+
+if hasattr(Binary, 'register_sniffable_binary_format'):
+    Binary.register_sniffable_binary_format('raw', 'raw', RAW)
+
+
+class Msp(Text):
+    """ Output of NIST MS Search Program chemdata.nist.gov/mass-spc/ftp/mass-spc/PepLib.pdf """
+    file_ext = "msp"
+
+    @staticmethod
+    def next_line_starts_with(contents, prefix):
+        next_line = contents.readline()
```

Comment:
I would like to avoid ``readline`` in sniffers - I have had it crash Python process before by reading in and attempting to allocate huge strings before randomly finding something that looks like a newline. How about the following variant of an MSP sniff that reads in a fixed size string.

```
with open(filename, 'r') as f:
     begin_contents = f.read(1024)
     if not "\n" in begin_contents:
          return False
     lines = begin_contents.splitlines()
     if len(lines) < 2:
           return False
     return lines[0].startswith("Name:") and lines[1].startswith("MW:")
```